Exception when have workaround because type erasure on JVM

Exception:

```java
C:\~\hibernate-orm\hibernate-core\src\main\java\org\hibernate\query\sqm\tree\domain\SqmTreatedRoot.java:107: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
	public Object accept(SemanticQueryWalker walker) {
	              ^
     Please report this at https://github.com/google/error-prone/issues/new and include the following:
  
     error-prone version: 2.41.0
     BugPattern: NullAway
     Stack Trace:
     java.lang.ClassCastException: class com.sun.tools.javac.code.Type$ForAll cannot be cast to class com.sun.tools.javac.code.Type$MethodType (com.sun.tools.javac.code.Type$ForAll and com.sun.tools.javac.code.Type$MethodType are in module jdk.compiler of loader 'app')
  	at com.uber.nullaway.generics.TypeSubstitutionUtils$RestoreNullnessAnnotationsVisitor.visitMethodType(TypeSubstitutionUtils.java:84)
  	at com.uber.nullaway.generics.TypeSubstitutionUtils$RestoreNullnessAnnotationsVisitor.visitMethodType(TypeSubstitutionUtils.java:69)
  	at jdk.compiler/com.sun.tools.javac.code.Type$MethodType.accept(Type.java:1495)
  	at jdk.compiler/com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4900)
  	at com.uber.nullaway.generics.TypeSubstitutionUtils.restoreExplicitNullabilityAnnotations(TypeSubstitutionUtils.java:61)
  	at com.uber.nullaway.generics.TypeSubstitutionUtils.memberType(TypeSubstitutionUtils.java:41)
  	at com.uber.nullaway.generics.GenericsChecks.checkTypeParameterNullnessForMethodOverriding(GenericsChecks.java:868)
  	at com.uber.nullaway.NullAway.matchMethod(NullAway.java:704)
  	at com.google.errorprone.scanner.ErrorProneScanner.processMatchers(ErrorProneScanner.java:509)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:817)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:170)
  	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:948)
  	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
  	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
  	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
  	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:119)
  	at jdk.compiler/com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:203)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:619)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:170)
  	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:855)
  	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:92)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
  	at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
  	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:119)
  	at jdk.compiler/com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:152)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:631)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:170)
  	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:623)
  	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:66)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:58)
  	at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
  	at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:231)
  	at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:133)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1436)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1383)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:963)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.invocationHelper(JavacTaskImpl.java:152)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)
  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)
  	at org.gradle.internal.compiler.java.IncrementalCompileTask.call(IncrementalCompileTask.java:92)
  	at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call(AnnotationProcessingCompileTask.java:94)
  	at org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call(ResourceCleaningCompilationTask.java:57)
  	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:83)
  	at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:50)
  	at org.gradle.api.internal.tasks.compile.daemon.AbstractIsolatedCompilerWorkerExecutor$CompilerWorkAction.execute(AbstractIsolatedCompilerWorkerExecutor.java:78)
  	at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
  	at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:54)
  	at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:48)
  	at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:100)
  	at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:48)
  	at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:32)
  	at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:22)
  	at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:108)
  	at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:77)
  	at org.gradle.process.internal.worker.request.WorkerAction.lambda$run$1(WorkerAction.java:150)
  	at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
  	at org.gradle.process.internal.worker.request.WorkerAction.lambda$run$2(WorkerAction.java:150)
  	at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:85)
  	at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:142)
  	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
  	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
  	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
  	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
  	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
  	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
  	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
  	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:48)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  	at java.base/java.lang.Thread.run(Thread.java:1583)
```

[Problematic code](https://github.com/hibernate/hibernate-orm/blob/b7591a94ecb7a34e87e708242ea871777dc6b1bf/hibernate-core/src/main/java/org/hibernate/query/sqm/tree/domain/SqmTreatedRoot.java#L105-L109):

```java
public class SqmTreatedRoot extends SqmRoot implements SqmTreatedFrom {
	@SuppressWarnings("unchecked")
	@Override
	public Object accept(SemanticQueryWalker walker) {
		return walker.visitTreatedPath( this );
	}
}
```

Java force to do with Object and raw types of parameter because _'accept(SemanticQueryWalker<T>)' in 'org.hibernate.query.sqm.tree.domain.SqmTreatedRoot' clashes with 'accept(SemanticQueryWalker<X>)' in 'org.hibernate.query.sqm.tree.from.SqmRoot'; both methods have same erasure, yet neither overrides the other_

In [SqmRoot](https://github.com/hibernate/hibernate-orm/blob/b7591a94ecb7a34e87e708242ea871777dc6b1bf/hibernate-core/src/main/java/org/hibernate/query/sqm/tree/from/SqmRoot.java#L161-L164) is:
```java
public class SqmRoot<E> extends AbstractSqmFrom<E,E> implements JpaRoot<E> {
	@Override
	public <X> X accept(SemanticQueryWalker<X> walker) {
		return walker.visitRootPath( this );
	}
}
```

In [SqmVisitableNode](https://github.com/hibernate/hibernate-orm/blob/b7591a94ecb7a34e87e708242ea871777dc6b1bf/hibernate-core/src/main/java/org/hibernate/query/sqm/tree/SqmVisitableNode.java#L19) is:

```java
public interface SqmVisitableNode extends SqmNode {
	/**
	 * Accept the walker per visitation
	 */
	<X> X accept(SemanticQueryWalker<X> walker);
}
```

Config:
```groovy
plugins {
    //id "org.checkerframework"
    id "net.ltgt.errorprone"
    id "net.ltgt.nullaway"
    id "de.thetaphi.forbiddenapis"
    id "com.diffplug.spotless"
}

dependencies {
    //compileOnly libs.checkerFramework
    //testCompileOnly libs.checkerFramework
    //checkerFramework libs.checkerFrameworkChecker
    api "org.jspecify:jspecify:1.0.0"
    // for use contract like @Nullable Boolean foo(final String a) @Contract(null->null)
    // see https://github.com/uber/NullAway/wiki/Supported-Annotations#contracts
    // compileOnly "org.jetbrains:annotations:26.0.2"
    errorprone "com.uber.nullaway:nullaway:0.12.9"
    errorprone "com.google.errorprone:error_prone_core:2.41.0"
}

tasks.withType(JavaCompile).configureEach {
    if (gradle.startParameter.taskNames.contains("ciCheck")) {
        // options.errorprone.enabled = true
    }
    if (it.name.toLowerCase().contains("test")) {
        options.errorprone.enabled = false
    }
    options.compilerArgs << "-Xmaxerrs" << "10000"
    options.errorprone {
        disableWarningsInGeneratedCode.set(true)

        // error prone core = warnings
        allErrorsAsWarnings.set(true)

        nullaway {
            error()
            // annotatedPackages.add("org.hibernate")
            onlyNullMarked.set(true)
            treatGeneratedAsUnannotated.set(true)
            checkContracts.set(true)
            checkOptionalEmptiness.set(true)
            exhaustiveOverride.set(true)
            jspecifyMode.set(true)
            customNullableAnnotations.add("org.checkerframework.checker.nullness.qual.Nullable")
        }
    }
}
```

How to handle this "unique" case?

_Aclaration_: This is an experiment branch with nullaway and jspecify, the only change is put `@NullMarked` on package-info

--- a/nullaway/src/main/java/com/uber/nullaway/generics/TypeSubstitutionUtils.java
+++ b/nullaway/src/main/java/com/uber/nullaway/generics/TypeSubstitutionUtils.java
@@ -81,6 +81,7 @@ public class TypeSubstitutionUtils {
 
     @Override
     public Type visitMethodType(Type.MethodType t, Type other) {
+      // other can be a ForAll whose qtype is a MethodType; asMethodType() safely unwraps it.
       Type.MethodType otherMethodType = other.asMethodType();
       List<Type> argtypes = t.argtypes;
       Type restype = t.restype;

--- a/nullaway/src/main/java/com/uber/nullaway/generics/TypeSubstitutionUtils.java
+++ b/nullaway/src/main/java/com/uber/nullaway/generics/TypeSubstitutionUtils.java
@@ -81,7 +81,7 @@ public class TypeSubstitutionUtils {
 
     @Override
     public Type visitMethodType(Type.MethodType t, Type other) {
-      Type.MethodType otherMethodType = (Type.MethodType) other;
+      Type.MethodType otherMethodType = other.asMethodType();
       List<Type> argtypes = t.argtypes;
       Type restype = t.restype;
       List<Type> thrown = t.thrown;

--- a/nullaway/src/test/java/com/uber/nullaway/jspecify/GenericInheritanceTests.java
+++ b/nullaway/src/test/java/com/uber/nullaway/jspecify/GenericInheritanceTests.java
@@ -207,6 +207,7 @@ public class GenericInheritanceTests extends NullAwayTestsBase {
         .doTest();
   }
 
+  /** For issue 1250. Checks that we don't crash when overriding with raw types. */
   @Test
   public void overrideWithRawTypes() {
     makeHelper()

--- a/nullaway/src/test/java/com/uber/nullaway/jspecify/GenericInheritanceTests.java
+++ b/nullaway/src/test/java/com/uber/nullaway/jspecify/GenericInheritanceTests.java
@@ -207,7 +207,7 @@ public class GenericInheritanceTests extends NullAwayTestsBase {
         .doTest();
   }
 
-  /** For issue 1250. Checks that we don't crash when overriding with raw types. */
+  /** For issue 1264. Checks that we don't crash when overriding with raw types. */
   @Test
   public void overrideWithRawTypes() {
     makeHelper()

--- a/nullaway/src/test/java/com/uber/nullaway/jspecify/GenericInheritanceTests.java
+++ b/nullaway/src/test/java/com/uber/nullaway/jspecify/GenericInheritanceTests.java
@@ -207,6 +207,30 @@ public class GenericInheritanceTests extends NullAwayTestsBase {
         .doTest();
   }
 
+  @Test
+  public void overrideWithRawTypes() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            "import org.jspecify.annotations.NullMarked;",
+            "@NullMarked",
+            "public class Test {",
+            "  static class A<T> {}",
+            "  static class B<T> {",
+            "    public <X> X accept(A<X> a) {",
+            "      throw new RuntimeException();",
+            "    }",
+            "  }",
+            "  static class Raw extends B {",
+            "    @Override",
+            "    public Object accept(A a) {",
+            "      throw new RuntimeException();",
+            "    }",
+            "  }",
+            "}")
+        .doTest();
+  }
+
   private CompilationTestHelper makeHelper() {
     return makeTestHelperWithArgs(
         Arrays.asList(

